src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -3429,6 +3437,22 @@ public DatabaseTopology ReadDatabaseTopology<TTransaction>(TransactionOperationC
             }
         }
 
+        public DatabaseTopology ReadDatabaseTopologyForShard<TTransaction>(TransactionOperationContext<TTransaction> context, string name, int shard)
+            where TTransaction : RavenTransaction
+        {
+            using (var databaseRecord = ReadRawDatabaseRecord(context, name))
+            {
+                if (databaseRecord.IsSharded() == false)
+                    throw new InvalidOperationException($"The database record '{name}' is not sharded.");
+
+                var topology = databaseRecord.Shards[shard];
```

Comment:
Check the length here, so we'll get good error and not just index out of range without knowing which db

Reply:
done